### ISSUE TITLE
[layernorm/rms_norm] FLOAT32 ROW_MAJOR gamma/beta is silently half-dropped on the non-distributed interleaved path

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#49902]
### Component / Area

ops / normalization (`ttnn.rms_norm`, `ttnn.layer_norm`)

### Issue Type

Bad Outputs

### Observed

`ttnn.rms_norm` / `ttnn.layer_norm` accept FLOAT32 gamma/beta in ROW_MAJOR layout — `validate` explicitly allows `FLOAT32 or BFLOAT16` (`layernorm_device_operation.cpp:110`) and the docstring advertises FLOAT32 — but silently return wrong results. No error, no warning.

On the non-distributed interleaved path, `reader_unary_interleaved_ln_rm_gb.cpp` hardcodes bf16 byte counts for the gamma/beta stick read:

| line (gamma / beta) | value | should be, for fp32 |
|---|---|---|
| L140 / L168 | read length `64` | `TILE_WIDTH * datum` = 128 |
| L147 / L175 | L1→L1 copy length `32` | `FACE_WIDTH * datum` = 64 |
| L151 / L179 | dest face offset `512` | `FACE_HW * datum` = 1024 |

Only channels 0–15 of each 32-channel tile are ever fetched from DRAM. Channels 16–31 live in face 1 — at byte offset 1024 for fp32, but the kernel writes to 512 — so they are **never written**, and read back whatever is left in L1.

Note the host already computes the correct stick size (`gamma_stick_size = padded_shape[-1] * element_size()`, `layernorm_op_multi_core.cpp:366`) and passes it as a compile-time arg, so page addressing is correct. Only the kernel's read *length* and face *offset* are wrong.

**Probe** — `x = ones`, so `rms_norm(x) == 1` and each output row *is* the applied gamma. With `gamma[i] = i`, every value identifies where it came from:

```
=== effective gamma applied, fp32 x ROW_MAJOR, gamma[i]=i, W=128 (Wt=4) ===
  tile 0: ok=16/32  [0..15,    0, 1, 1, -1, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0]
  tile 1: ok=16/32  [32..47,   0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]
  tile 2: ok=16/32  [64..79,   0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  tile 3: ok=16/32  [96..111,  0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, -1, 0]
  TOTAL correct: 64/128
```

Exactly **16/32 per tile, at every width** — verified W=64 → 32/64, W=128 → 64/128, W=256 → 128/256.

**PCC:**

```
=== ttnn.rms_norm, W=128, bf16 input ===
  gamma bf16 x ROW_MAJOR -> PCC 0.999976    ok
  gamma fp32 x ROW_MAJOR -> PCC 0.314757    <-- BUG
  gamma bf16 x TILE      -> PCC 0.999976    ok
  gamma fp32 x TILE      -> PCC 0.999975    ok  (TILE gamma uses a different, tile-granular reader)

=== ttnn.layer_norm, W=128, bf16 input ===
  gamma/beta bf16 x ROW_MAJOR -> PCC 0.999992   ok
  gamma/beta fp32 x ROW_MAJOR -> PCC 0.180787   <-- BUG (worse: beta is additive)
```

The dropped channels read **uninitialized L1**, so the PCC is not stable: the same fp32×RM `rms_norm` call gives **0.700646** in a fresh process (residue is zeros → exactly √0.5, half the channels lost) but **0.314757** when another program ran first (residue is junk). The deterministic invariant is the 16/32-per-tile count above. This matches #48226 describing the same bug in the sibling reader as "read garbage (NaN)".

### Expected

Either produce correct results for FLOAT32 ROW_MAJOR gamma/beta as `validate` and the docs promise, or reject the combination with a `TT_FATAL`. Silently applying half the weights is the worst of the three outcomes.

### 1. Steps (exact commands)

```bash
cd $TT_METAL_HOME && source python_env/bin/activate
python repro_fp32_rm_gamma.py
```

<details><summary>repro_fp32_rm_gamma.py (self-contained)</summary>

```python
import torch, ttnn

W, S, EPS, TW = 128, 32, 1e-5, 32

def pcc(a, b):
    a, b = a.flatten().float(), b.flatten().float()
    return torch.corrcoef(torch.stack([a, b]))[0, 1].item()

def to_rm_gamma(t, dtype, device):   # ROW_MAJOR gamma/beta must be [1, 1, W/32, 32]
    return ttnn.from_torch(t.reshape(1, 1, W // TW, TW), dtype=dtype, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)

def to_tile_gamma(t, dtype, device):
    return ttnn.from_torch(t.reshape(1, 1, 1, W), dtype=dtype, layout=ttnn.TILE_LAYOUT, device=device)

device = ttnn.open_device(device_id=0)
try:
    # 1. Which gamma does the kernel actually apply?
    #    x = ones => rms_norm(x) == 1 => each output row IS the applied gamma.
    g_ar = torch.arange(W, dtype=torch.float32)
    xo = torch.ones((1, S, W), dtype=torch.float32)
    out = ttnn.rms_norm(ttnn.from_torch(xo, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device),
                        epsilon=EPS, weight=to_rm_gamma(g_ar, ttnn.float32, device))
    eff = ttnn.to_torch(out).float().reshape(xo.shape)[0, 0]
    print(f"\n=== effective gamma applied, fp32 x ROW_MAJOR, gamma[i]=i, W={W} (Wt={W // TW}) ===")
    for t in range(W // TW):
        sl = eff[t * TW:(t + 1) * TW]
        ok = int((sl - g_ar[t * TW:(t + 1) * TW]).abs().lt(0.5).sum())
        print(f"  tile {t}: ok={ok:2d}/32  {[int(v) for v in sl.tolist()]}")
    print(f"  TOTAL correct: {int((eff - g_ar).abs().lt(0.5).sum())}/{W}")

    # 2. PCC across gamma dtype x layout
    torch.manual_seed(0)
    x = torch.rand((1, S, W), dtype=torch.float32)
    g = torch.rand((W,), dtype=torch.float32)
    b = torch.rand((W,), dtype=torch.float32)
    xt = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

    print(f"\n=== ttnn.rms_norm, W={W}, bf16 input ===")
    ref_r = x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + EPS) * g
    for name, dt, mk in [("gamma bf16 x ROW_MAJOR", ttnn.bfloat16, to_rm_gamma),
                         ("gamma fp32 x ROW_MAJOR", ttnn.float32,  to_rm_gamma),
                         ("gamma bf16 x TILE     ", ttnn.bfloat16, to_tile_gamma),
                         ("gamma fp32 x TILE     ", ttnn.float32,  to_tile_gamma)]:
        got = ttnn.to_torch(ttnn.rms_norm(xt, epsilon=EPS, weight=mk(g, dt, device))).float().reshape(x.shape)
        print(f"  {name} -> PCC {pcc(ref_r, got):.6f}")

    print(f"\n=== ttnn.layer_norm, W={W}, bf16 input ===")
    mu = x.mean(-1, keepdim=True)
    ref_l = (x - mu) * torch.rsqrt(x.var(-1, unbiased=False, keepdim=True) + EPS) * g + b
    for name, dt in [("gamma/beta bf16 x ROW_MAJOR", ttnn.bfloat16),
                     ("gamma/beta fp32 x ROW_MAJOR", ttnn.float32)]:
        out = ttnn.layer_norm(xt, epsilon=EPS, weight=to_rm_gamma(g, dt, device), bias=to_rm_gamma(b, dt, device))
        print(f"  {name} -> PCC {pcc(ref_l, ttnn.to_torch(out).float().reshape(x.shape)):.6f}")
finally:
    ttnn.close_device(device)
```
</details>

### 2. Input data / link or description

None. Self-contained — inputs are `torch.rand` / `torch.arange`. No model weights or datasets.

### 3. Frequency

Always (100%). The set of corrupted channels is deterministic — channels 16–31 of every tile, at every width. The *values* landing in those channels come from uninitialized L1, so the resulting PCC varies with run order.

### 1. Software Versions

```
OS:               Ubuntu 24.04.3 LTS (Noble Numbat)
tt-metal:         built from source @ 6eeee2553 (2026-07-13)
                  also present on main @ 1cb6f8e40e (2026-07-15)
Python:           3.12.12
torch:            2.11.0+cpu
KMD:              2.7.0
Firmware bundle:  19.7.0
```

### 2. Hardware Details

```
Product:      Blackhole
Card/System:  P150 x4 (QuietBox); repro uses device 0 only
```

Not hardware-specific — this is a pure dataflow transfer-size bug.

### Is this a regression?

No.

### Regression Details

Not a regression. The gamma/beta stick read has **never** been datum-size-aware:

- the original 2023 implementation hardcoded `32`-byte reads;
- **#15993** (2024-12-13, "DRAM Read Alignment for Layernorm", ticket #14353) restructured them to `64` + `32` for DRAM read alignment, preserving the bf16 assumption;
- the Device 2.0 port **#39453** (2026-03-09) carried them over verbatim.

Broken since the ROW_MAJOR gamma path was introduced.

### Logs & Diagnostics

No error output — the silence is the bug.

### Priority

P2

### Impact

Any model passing fp32 norm weights in ROW_MAJOR layout gets silently degraded output. Hit during Qwen3.5 bring-up on Blackhole. **Workaround:** cast gamma/beta to `bfloat16`, or use TILE layout.

### Notes — relation to existing work, and suggested fix

Two of the three gamma-stick readers in the tree already handle this correctly; the non-distributed interleaved one is the only one left:

| variant | file | fp32 RM gamma |
|---|---|---|
| distributed post-AG | `..._rm_gb_post_allgather.cpp` | ✅ fixed by **#48226** (2026-07-08), now `datum_bytes`-aware |
| non-dist **sharded** | `writer_unary_sharded_ln_rm_gb.cpp:95-96` | ✅ already correct via `FLOAT32_DTYPE_GAMMA ? 64 : 32` / `? 1024 : 512` |
| non-dist **interleaved** | `reader_unary_interleaved_ln_rm_gb.cpp:140,147,151` | ❌ **hardcoded 64 / 32 / 512** |

The hardcoded `64/32/512` are exactly the `FLOAT32_DTYPE_GAMMA == false` branch of the sharded kernel, so the fix should be a small port of either sibling.

This file has prior form as the odd one out: **#43913** (fixed by #47215, 2026-06-18) was a different bug in this same reader, and its write-up concluded *"Symmetry — the non-rm_gb reader already does it; rm_gb is the outlier."* The same asymmetry applies here, now on the dtype axis.

**Why it was missed:** the reader is selected by gamma *layout* (`use_row_major_kernel`, `layernorm_op_multi_core.cpp:216`). `test_rms_norm_row_major` — the only RM-gamma test — is bf16-only, and `test_layernorm_interleaved_all_config` (added by #48226) parametrizes `gamma_dtype` over fp32/bf16 but builds gamma with `TILE_LAYOUT`. fp32 × ROW_MAJOR gamma is the empty cell.

**Checked against existing issues — believed distinct from all of:**

| # | why it's different |
|---|---|
| #44650 | scoped to LayerNormPostAllGather + GroupNorm; remaining work per the thread is GroupNorm and FP32-end-to-end / `UnpackToDestFp32`. That's a *compute precision* axis (TF32 vs FP32); this is a *dataflow transfer-size* bug — making the pipeline FP32 end-to-end would not move the PCC, since half the weights never arrive. The stage tables there also have no row for the gamma/beta read (there is one for the input read); the gamma rows start from `cb_gamma` already being correct |
| #43946 | hardcoded `Float16_b` **intermediate CB formats**, host-side, in distributed pre-AG program factories |
| #49600 | fp32 **input** precision (max abs err ~5.5e-3), TILE layout, gamma not involved |
| #48824 | `fp32_dest_acc_en=False` PCC drop — compute config |
| #38561 | `fp32_dest_acc_en` on BH with bf16 gamma — compute config |
| #48220 | gamma **shape** constraint (`TT_FATAL`) — loud, and about shape not dtype |
| #43913 | same file, different bug (missing partial scaler tile → hang); fixed by #47215 |
| #37551 | groupnorm *rejects* non-bf16 gamma — loud, not silent |

Happy to put up a PR.

**Comment by kinginu** [comment]
Opened #49974 with a fix.

It makes the gamma/beta stick read in `reader_unary_interleaved_ln_rm_gb.cpp` datum-size-aware, derived from the CB tile size the same way the post-all-gather sibling does — so no host change and no new compile-time arg. On Blackhole, `rms_norm` with fp32 ROW_MAJOR gamma goes from PCC 0.31 to 0.999975 and `layer_norm` from 0.18 to 0.999994, with the bf16 paths unchanged.

The PR also adds a `gamma_layout` axis to `test_layernorm_interleaved_all_config`, which is the matrix cell that was missing. Two notes for whoever picks this up:

- Adding that axis exposed **two pre-existing device hangs** involving ROW_MAJOR gamma/beta — (A) ROW_MAJOR input + ROW_MAJOR gamma/beta, for any input dtype, and (B) Welford + fp32 input + ROW_MAJOR gamma/beta. Both reproduce on unpatched `main` and are unaffected by this fix, so the PR skips them and I will file them separately.
- `tests/ttnn/nightly/` only runs on `schedule`/`workflow_dispatch`, never on `pull_request`, so the new cells will not run on the PR itself. All results are from local Blackhole runs (26 passed / 22 skipped / 0 failed, plus 355 passed in the existing rms_norm + layer_norm suites).
